src/transformers/models/modernbert_decoder/modular_modernbert_decoder.py

```diff
@@ -0,0 +1,1305 @@
+        key: Key tensor of shape [batch_size, num_heads, seq_len, head_dim]
+        value: Value tensor of shape [batch_size, num_heads, seq_len, head_dim]
+        attention_mask: Attention mask tensor (should already incorporate causal and sliding window constraints)
+        dropout: Dropout probability
+        scaling: Attention scaling factor (will use module.scaling if None)
+        sliding_window: Sliding window size (for local attention) - not used in eager, mask should already be prepared
+        is_causal: Whether to use causal attention
+
+    Returns:
+        Tuple of (attention_output, attention_weights)
+    """
+    batch_size, num_heads, seq_len, head_dim = query.shape
+    kv_seq_len = key.shape[-2]
+
+    # Use consistent scaling like the encoder version
+    scale = module.head_dim**-0.5
+
+    # Compute attention scores
+    attn_weights = torch.matmul(query, key.transpose(-2, -1)) * scale
+
+    if is_causal:
+        past_length = kv_seq_len - seq_len
+
+        causal_mask = torch.full(
+            (seq_len, kv_seq_len),
+            torch.finfo(attn_weights.dtype).min,
+            device=attn_weights.device,
+            dtype=attn_weights.dtype,
+        )
+
+        # Create indices for the causal pattern
+        row_indices = torch.arange(seq_len, device=attn_weights.device).unsqueeze(1)
+        col_indices = torch.arange(kv_seq_len, device=attn_weights.device).unsqueeze(0)
+
+        # Allow attention to past positions + current and previous positions in the current sequence
+        causal_pattern = col_indices <= (past_length + row_indices)
+        causal_mask = causal_mask.masked_fill(causal_pattern, 0)
+
+        # Apply sliding window constraint if specified
+        if sliding_window is not None and sliding_window > 0:
+            # Create sliding window mask - limit attention to within sliding window distance
+            # Calculate absolute distance between query positions and key positions
+            query_positions = past_length + row_indices  # Actual positions in the full sequence
+            key_positions = col_indices  # Key positions in the full sequence
+            distance = torch.abs(query_positions - key_positions)
+
+            # Create sliding window pattern (allow attention within window)
+            window_pattern = distance <= sliding_window
+
+            # Combine causal and sliding window constraints - both must be satisfied
+            combined_pattern = causal_pattern & window_pattern
+
+            # Reset mask and apply combined constraints
+            causal_mask = torch.full(
+                (seq_len, kv_seq_len),
+                torch.finfo(attn_weights.dtype).min,
+                device=attn_weights.device,
+                dtype=attn_weights.dtype,
+            )
+            causal_mask = causal_mask.masked_fill(combined_pattern, 0)
+
+        attn_weights = attn_weights + causal_mask
```

Comment:
There should be 0 code path here. Causal or not causal only depends on the mask, which you are suppose to only create and compute once.